model/riscv_insts_cmoext.sail

```diff
@@ -0,0 +1,245 @@
+/* ****************************************************************** */
+/* This file specifies the instructions in the cmo extension */
+
+
+/* ****************************************************************** */
+union clause ast = RISCV_CBO : (cbop, regidx)
+
+mapping encdec_cbop : cbop <-> bits(12) = {
+  CBO_CLEAN <-> 0b000000000001,
+  CBO_FLUSH <-> 0b000000000010,
+  CBO_INVAL <-> 0b000000000000,
+  CBO_ZERO  <-> 0b000000000100
+}
+
+mapping clause encdec = RISCV_CBO(cbop, rs1)
+  <-> encdec_cbop(cbop) @ rs1 @ 0b010 @ 0b00000 @ 0b0001111
+
+mapping cbo_mnemonic : cbop <-> string = {
+  CBO_CLEAN <-> "cbo.clean",
+  CBO_FLUSH <-> "cbo.flush",
+  CBO_INVAL <-> "cbo.inval",
+  CBO_ZERO  <-> "cbo.zero"
+}
+
+mapping clause assembly = RISCV_CBO(cbop, rs1)
+  <-> cbo_mnemonic(cbop) ^ spc() ^ reg_name(rs1)
+
+val write_zero: (xlenbits, xlenbits) -> MemoryOpResult(bool) effect {wmv, wmvt, rreg, wreg, escape}
+function write_zero(paddr, width) = {
+  let res : MemoryOpResult(bool) = mem_write_value(paddr, 8, to_bits(64, 0), false, false, false);
+  match (res) {
+    MemValue(true)  => {
+      if(unsigned(width) > 8) then
+        write_zero(paddr + 8, width - 8)
+      else
+        MemValue(true)
+    },
+    MemValue(false) => MemValue(false),
+    MemException(e) => MemException(e)
+  }
+}
+
+val process_cbo_zero : (xlenbits, xlenbits) -> Retired effect {eamem, escape, rmem, rmemt, rreg, wmv, wmvt, wreg}
+function process_cbo_zero(vaddr, width)  = {
+  match ext_data_get_addr(to_bits(5, 0), vaddr, Write(Data), BYTE) {
+    Ext_DataAddr_Error(e)  => { ext_handle_data_check_error(e); RETIRE_FAIL },
+    Ext_DataAddr_OK(vaddr) =>
+      match translateAddr(vaddr, Write(Data)) {
+        TR_Failure(e, _)    => { handle_mem_exception(vaddr, e); RETIRE_FAIL },
+        TR_Address(paddr, _) => {
+          let eares : MemoryOpResult(unit) = mem_write_ea(paddr, 1, false, false, false);
+          match (eares) {
+            MemException(e) => { handle_mem_exception(vaddr, e); RETIRE_FAIL },
+            MemValue(_) => {
+              let res : MemoryOpResult(bool) = write_zero(paddr, width);
+              match (res) {
+                MemValue(true)  => RETIRE_SUCCESS,
+                MemValue(false) => internal_error("store got false from mem_write_value"),
+                MemException(e) => { handle_mem_exception(vaddr, e); RETIRE_FAIL }
+              }
+            }
+          }
+        }
+      }
+  }
+}
+
+union translation_check = {
+  TransTrap : ExceptionType,
+  TransSuccess : unit
+}
+
+val process_clean_inval : (xlenbits, xlenbits, bool, bool) -> Retired effect {escape, rmem, rmemt, rreg, wmv, wmvt, wreg}
+function process_clean_inval(vaddr, width, clean, inval) = {
+  let load: translation_check = match ext_data_get_addr(to_bits(5, 0), vaddr, Read(Data), BYTE) {
+    Ext_DataAddr_Error(e)  => TransTrap(E_Reserved_14()),
+    Ext_DataAddr_OK(vaddr) =>
+      match translateAddr(vaddr, Read(Data)) {
```

Comment:
How does invalidate interact with the dirty bit of page table entries? Just translating a read won't dirty the page, but surely an invalidate needs to dirty the page if actually invalidating rather than just cleaning/flushing? Though whether it's "only dirty if the cache line is dirty" or "always dirty even if there's nothing to write back" I don't know. I can only see discussion of accessed and dirty bits for prefetch and zeroing, not management instruction.

Reply:
(The answer to that affects whether or not you can clean this up by just hoisting all three load, store and fetch calculations and matching on them all at the same time, which would flatten this function somewhat)